# Your Title Here

**Name(s)**: Nessa Pantfoerder

**Website Link**: https://nessapantfoerder.github.io/power_outage_analysis/

## Code

In [3]:
import pandas as pd
import numpy as np
import os

import plotly.express as px
pd.options.plotting.backend = 'plotly'

### Cleaning and EDA

**GENERAL INFORMATION**

*Time of the outage event*
- YEAR	
    - Indicates the year when the outage event occurred
- MONTH
    - Indicates the month when the outage event occurred
    
*Geographic areas*
- U.S._STATE
    - Represents all the states in the continental U.S.
- POSTAL.CODE
    - Represents the postal code of the U.S. states
- NERC.REGION
    - The North American Electric Reliability Corporation (NERC) regions involved in the outage event

**REGIONAL CLIMATE INFORMATION**

*U.S. Climate regions*
- CLIMATE.REGION
    - U.S. Climate regions as specified by National Centers for Environmental Information (nine climatically consistent regions in continental U.S.A.)

*El Niño/La Niña*
- ANOMALY.LEVEL
    - This represents the oceanic El Niño/La Niña (ONI) index referring to the cold and warm episodes by season. It is estimated as a 3-month running mean of ERSST.v4 SST anomalies in the Niño 3.4 region (5°N to 5°S, 120–170°W) [6]
- CLIMATE.CATEGORY
    - This represents the climate episodes corresponding to the years. The categories—“Warm”, “Cold” or “Normal” episodes of the climate are based on a threshold of ± 0.5 °C for the Oceanic Niño Index (ONI)

**OUTAGE EVENTS INFORMATION**

*Event start and end information*
- OUTAGE.START.DATE
    - This variable indicates the day of the year when the outage event started (as reported by the corresponding Utility in the region)
- OUTAGE.START.TIME
    - This variable indicates the time of the day when the outage event started (as reported by the corresponding Utility in the region)
- OUTAGE.RESTORATION.DATE
    - This variable indicates the day of the year when power was restored to all the customers (as reported by the corresponding Utility in the region)
- OUTAGE.RESTORATION.TIME
    - This variable indicates the time of the day when power was restored to all the customers (as reported by the corresponding Utility in the region)

*Cause of the event*
- CAUSE.CATEGORY
    - Categories of all the events causing the major power outages
- CAUSE.CATEGORY.DETAIL
    - Detailed description of the event categories causing the major power outages
- HURRICANE.NAMES
    - If the outage is due to a hurricane, then the hurricane name is given by this variable
*Extent of outages*
- OUTAGE.DURATION	
    - Duration of outage events (in minutes)
- DEMAND.LOSS.MW
    - Amount of peak demand lost during an outage event (in Megawatt) [but in many cases, total demand is reported]
- CUSTOMERS.AFFECTED
    - Number of customers affected by the power outage event

**REGIONAL ELECTRICITY CONSUMPTION INFORMATION**

*Electricity price*	
- RES.PRICE
    - Monthly electricity price in the residential sector (cents/kilowatt-hour)
- COM.PRICE
    - Monthly electricity price in the commercial sector (cents/kilowatt-hour)
- IND.PRICE	
    - Monthly electricity price in the industrial sector (cents/kilowatt-hour)
- TOTAL.PRICE
    - Average monthly electricity price in the U.S. state (cents/kilowatt-hour)

*Electricity consumption*
- RES.SALES	
    - Electricity consumption in the residential sector (megawatt-hour)
- COM.SALES
    - Electricity consumption in the commercial sector (megawatt-hour)
- IND.SALES	
    - Electricity consumption in the industrial sector (megawatt-hour)
- TOTAL.SALES
    - Total electricity consumption in the U.S. state (megawatt-hour)
- RES.PERCEN
    - Percentage of residential electricity consumption compared to the total electricity consumption in the state (in %)
- COM.PERCEN
    - Percentage of commercial electricity consumption compared to the total electricity consumption in the state (in %)
- IND.PERCEN
    - Percentage of industrial electricity consumption compared to the total electricity consumption in the state (in %)

*Customers served*
- RES.CUSTOMERS
    - Annual number of customers served in the residential electricity sector of the U.S. state
- COM.CUSTOMERS
    - Annual number of customers served in the commercial electricity sector of the U.S. state
- IND.CUSTOMERS
    - Annual number of customers served in the industrial electricity sector of the U.S. state
- TOTAL.CUSTOMERS
    - Annual number of total customers served in the U.S. state
- RES.CUST.PCT
    - Percent of residential customers served in the U.S. state (in %)
- COM.CUST.PCT
    - Percent of commercial customers served in the U.S. state (in %)
- IND.CUST.PCT
    - Percent of industrial customers served in the U.S. state (in %)

**REGIONAL ECONOMIC CHARACTERISTICS**

*Economic outputs*
- PC.REALGSP.STATE
    - Per capita real gross state product (GSP) in the U.S. state (measured in 2009 chained U.S. dollars)
- PC.REALGSP.USA
    - Per capita real GSP in the U.S. (measured in 2009 chained U.S. dollars)
- PC.REALGSP.REL
    - Relative per capita real GSP as compared to the total per capita real GDP of the U.S. (expressed as fraction of per capita State real GDP & per capita US real GDP)
- PC.REALGSP.CHANGE
    - Percentage change of per capita real GSP from the previous year (in %)
- UTIL.REALGSP
    - Real GSP contributed by Utility industry (measured in 2009 chained U.S. dollars)
- TOTAL.REALGSP
    - Real GSP contributed by all industries (total) (measured in 2009 chained U.S. dollars)
- UTIL.CONTRI
    - Utility industry׳s contribution to the total GSP in the State (expressed as percent of the total real GDP that is contributed by the Utility industry) (in %)
- PI.UTIL.OFUSA
    - State utility sector׳s income (earnings) as a percentage of the total earnings of the U.S. utility sector׳s income (in %)

**REGIONAL LAND-USE CHARACTERICS**

*Population*
- POPULATION
    - Population in the U.S. state in a year
- POPPCT_URBAN
    - Percentage of the total population of the U.S. state represented by the urban population (in %)
- POPPCT_UC
    - Percentage of the total population of the U.S. state represented by the population of the urban clusters (in %)
- POPDEN_URBAN
    - Population density of the urban areas (persons per square mile)
- POPDEN_UC
    - Population density of the urban clusters (persons per square mile)
- POPDEN_RURAL
    - Population density of the rural areas (persons per square mile)
    
*Land area*
- AREAPCT_URBAN
    - Percentage of the land area of the U.S. state represented by the land area of the urban areas (in %)
- AREAPCT_UC
    - Percentage of the land area of the U.S. state represented by the land area of the urban clusters (in %)
- PCT_LAND
    - Percentage of land area in the U.S. state as compared to the overall land area in the continental U.S. (in %)
- PCT_WATER_TOT
    - Percentage of water area in the U.S. state as compared to the overall water area in the continental U.S. (in %)
- PCT_WATER_INLAND
    - Percentage of inland water area in the U.S. state as compared to the overall inland water area in the continental U.S. (in %)

In [4]:
outage = pd.read_csv('outage.csv')
print(outage.shape[0])
outage.head()

1534


OBS  YEAR  MONTH U.S._STATE POSTAL.CODE NERC.REGION      CLIMATE.REGION  \
0    1  2011    7.0  Minnesota          MN         MRO  East North Central   
1    2  2014    5.0  Minnesota          MN         MRO  East North Central   
2    3  2010   10.0  Minnesota          MN         MRO  East North Central   
3    4  2012    6.0  Minnesota          MN         MRO  East North Central   
4    5  2015    7.0  Minnesota          MN         MRO  East North Central   

   ANOMALY.LEVEL CLIMATE.CATEGORY          OUTAGE.START.DATE  ...  \
0           -0.3           normal      Friday, July 01, 2011  ...   
1           -0.1           normal       Sunday, May 11, 2014  ...   
2           -1.5             cold  Tuesday, October 26, 2010  ...   
3           -0.1           normal     Tuesday, June 19, 2012  ...   
4            1.2             warm    Saturday, July 18, 2015  ...   

  POPPCT_URBAN POPPCT_UC POPDEN_URBAN POPDEN_UC POPDEN_RURAL AREAPCT_URBAN  \
0        73.27     15.28       2279.0    1700.5         18.2          2.14   
1        73.27     15.28       2279.0    1700.5         18.2          2.14   
2        73.27     15.28       2279.0    1700.5         18.2          2.14   
3        73.27     15.28       2279.0    1700.5         18.2          2.14   
4        73.27     15.28       2279.0    1700.5         18.2          2.14   

   AREAPCT_UC   PCT_LAND  PCT_WATER_TOT  PCT_WATER_INLAND  
0         0.6  91.592666       8.407334          5.478743  
1         0.6  91.592666       8.407334          5.478743  
2         0.6  91.592666       8.407334          5.478743  
3         0.6  91.592666       8.407334          5.478743  
4         0.6  91.592666       8.407334          5.478743  

[5 rows x 56 columns]

**Introduction:**

This dataset outlines major power outages in the continental U.S. from january 2000 to July 2016 providing information on population, region, land area, consumption, outage duration and causes, as well as regional climate information.

*Question:*

- Does the cause of major outages affect the duration of the outage?

*Rows:*

There are 1534 rows in this dataset.

*Relevant Columns and Descriptions:*

- YEAR	
    - Indicates the year when the outage event occurred
- MONTH
    - Indicates the month when the outage event occurred
- U.S._STATE
    - Represents all the states in the continental U.S.
- POSTAL.CODE
    - Represents the postal code of the U.S. states
- NERC.REGION
    - The North American Electric Reliability Corporation (NERC) regions involved in the outage event
- CLIMATE.REGION
    - U.S. Climate regions as specified by National Centers for Environmental Information (nine climatically consistent regions in continental U.S.A.)
- ANOMALY.LEVEL
    - This represents the oceanic El Niño/La Niña (ONI) index referring to the cold and warm episodes by season. It is estimated as a 3-month running mean of ERSST.v4 SST anomalies in the Niño 3.4 region (5°N to 5°S, 120–170°W) [6]
- CLIMATE.CATEGORY
    - This represents the climate episodes corresponding to the years. The categories—“Warm”, “Cold” or “Normal” episodes of the climate are based on a threshold of ± 0.5 °C for the Oceanic Niño Index (ONI)
- OUTAGE.START.DATE
    - This variable indicates the day of the year when the outage event started (as reported by the corresponding Utility in the region)
- OUTAGE.START.TIME
    - This variable indicates the time of the day when the outage event started (as reported by the corresponding Utility in the region)
- OUTAGE.RESTORATION.DATE
    - This variable indicates the day of the year when power was restored to all the customers (as reported by the corresponding Utility in the region)
- OUTAGE.RESTORATION.TIME
    - This variable indicates the time of the day when power was restored to all the customers (as reported by the corresponding Utility in the region)
- OUTAGE.DURATION	
    - Duration of outage events (in minutes)
- DEMAND.LOSS.MW
    - Amount of peak demand lost during an outage event (in Megawatt) [but in many cases, total demand is reported]
- CUSTOMERS.AFFECTED
    - Number of customers affected by the power outage event
- POPULATION
    - Population in the U.S. state in a year
- POPPCT_URBAN
    - Percentage of the total population of the U.S. state represented by the urban population (in %)
- POPPCT_UC
    - Percentage of the total population of the U.S. state represented by the population of the urban clusters (in %)
- POPDEN_URBAN
    - Population density of the urban areas (persons per square mile)
- POPDEN_UC
    - Population density of the urban clusters (persons per square mile)
- POPDEN_RURAL
    - Population density of the rural areas (persons per square mile)
- AREAPCT_URBAN
    - Percentage of the land area of the U.S. state represented by the land area of the urban areas (in %)
- AREAPCT_UC
    - Percentage of the land area of the U.S. state represented by the land area of the urban clusters (in %)
- PCT_LAND
    - Percentage of land area in the U.S. state as compared to the overall land area in the continental U.S. (in %)
- PCT_WATER_TOT
    - Percentage of water area in the U.S. state as compared to the overall water area in the continental U.S. (in %)
- PCT_WATER_INLAND
    - Percentage of inland water area in the U.S. state as compared to the overall inland water area in the continental U.S. (in %)


In [5]:
outage_clean = outage.replace('NA', np.NaN)

#Combine date and time into one column
outage_clean['OUTAGE.START'] = pd.to_datetime(outage_clean['OUTAGE.START.DATE'].apply(str) + ' '+ outage_clean['OUTAGE.START.TIME'])
outage_clean['OUTAGE.RESTORATION'] = pd.to_datetime(outage_clean['OUTAGE.RESTORATION.DATE'].apply(str) +' '+ outage_clean['OUTAGE.RESTORATION.TIME'])

#drop date and time
outage_clean = outage_clean.drop(columns=['OUTAGE.START.DATE', 'OUTAGE.START.TIME', 'OUTAGE.RESTORATION.DATE', 'OUTAGE.RESTORATION.TIME'])

outage_clean.head()

OBS  YEAR  MONTH U.S._STATE POSTAL.CODE NERC.REGION      CLIMATE.REGION  \
0    1  2011    7.0  Minnesota          MN         MRO  East North Central   
1    2  2014    5.0  Minnesota          MN         MRO  East North Central   
2    3  2010   10.0  Minnesota          MN         MRO  East North Central   
3    4  2012    6.0  Minnesota          MN         MRO  East North Central   
4    5  2015    7.0  Minnesota          MN         MRO  East North Central   

   ANOMALY.LEVEL CLIMATE.CATEGORY      CAUSE.CATEGORY  ... POPDEN_URBAN  \
0           -0.3           normal      severe weather  ...       2279.0   
1           -0.1           normal  intentional attack  ...       2279.0   
2           -1.5             cold      severe weather  ...       2279.0   
3           -0.1           normal      severe weather  ...       2279.0   
4            1.2             warm      severe weather  ...       2279.0   

  POPDEN_UC  POPDEN_RURAL  AREAPCT_URBAN  AREAPCT_UC   PCT_LAND  \
0    1700.5          18.2           2.14         0.6  91.592666   
1    1700.5          18.2           2.14         0.6  91.592666   
2    1700.5          18.2           2.14         0.6  91.592666   
3    1700.5          18.2           2.14         0.6  91.592666   
4    1700.5          18.2           2.14         0.6  91.592666   

   PCT_WATER_TOT  PCT_WATER_INLAND        OUTAGE.START  OUTAGE.RESTORATION  
0       8.407334          5.478743 2011-07-01 17:00:00 2011-07-03 20:00:00  
1       8.407334          5.478743 2014-05-11 18:38:00 2014-05-11 18:39:00  
2       8.407334          5.478743 2010-10-26 20:00:00 2010-10-28 22:00:00  
3       8.407334          5.478743 2012-06-19 04:30:00 2012-06-20 23:00:00  
4       8.407334          5.478743 2015-07-18 02:00:00 2015-07-19 07:00:00  

[5 rows x 54 columns]

**Data Cleaning:**

I combined columns 'OUTAGE.START.DATE' and 'OUTAGE.START.TIME' into one column 'OUTAGE.START' of type pd.Timestamp

I did the same combining 'OUTAGE.RESTORATION.DATE' and 'OUTAGE.RESTORATION.TIME' into 'OUTAGE.RESTORATION'

I then replaced all missing 'NA' values with np.NaN

In [6]:
fig = px.histogram(outage_clean, x="OUTAGE.DURATION", histnorm='probability',
                   title="Duration of Outage in Minutes")
fig.show()

In [7]:
fig = px.histogram(outage_clean, x="CUSTOMERS.AFFECTED", histnorm='probability',
                   title="Customers Affected by Outages")
fig.show()

In [ ]:
fig.write_html('univariate.html', include_plotlyjs='cdn')

**Univariate Analysis:**

Customers affected by Outages shows that it is more likely that fewer customers will be affected. Meaning that an outage that affects the  maximum amount of customers is much rarer than outages that affect the minimum amount of cutsomers.

In [8]:
fig = px.scatter(outage_clean, x="CUSTOMERS.AFFECTED", y="OUTAGE.DURATION", 
                    title='Customers Affected by Outage Duration in Minutes')
fig.show()

In [ ]:
fig.write_html('bivariate.html', include_plotlyjs='cdn')

In [9]:
fig = px.box(outage_clean, x="OUTAGE.DURATION", y="CLIMATE.REGION", title= 'Outage Duration in Minutes by Climate Region')
fig.show()

**Bivariate Analysis:**

Customers Affected by Outage Duration in Minutes shows that the majority of outages are short in duration in minimal in customers affected. Outliers that are greater in customers affected tend to also be greater in duration with a few short outages.

In [10]:
outage_clean.pivot_table(index='CLIMATE.REGION', columns='CLIMATE.CATEGORY', aggfunc='size')


CLIMATE.CATEGORY    cold  normal  warm
CLIMATE.REGION                        
Central               68     102    29
East North Central    38      83    17
Northeast            124     180    45
Northwest             47      49    36
South                 57     111    59
Southeast             45      71    34
Southwest             24      46    22
West                  64      93    60
West North Central     5       7     4

In [11]:
outage_clean.pivot_table(index='CAUSE.CATEGORY', columns='MONTH', aggfunc='size')

MONTH                          1.0   2.0   3.0   4.0   5.0    6.0   7.0   \
CAUSE.CATEGORY                                                             
equipment failure               3.0   4.0   4.0   7.0   6.0    7.0  12.0   
fuel supply emergency           4.0   9.0   5.0   3.0   2.0    7.0   6.0   
intentional attack             50.0  40.0  42.0  43.0  46.0   34.0  34.0   
islanding                       NaN   6.0   3.0   2.0   5.0   10.0   4.0   
public appeal                   4.0   5.0   2.0   NaN   5.0   16.0  14.0   
severe weather                 67.0  61.0  30.0  44.0  51.0  102.0  98.0   
system operability disruption   8.0  11.0  14.0  12.0  12.0   19.0  13.0   

MONTH                          8.0   9.0   10.0  11.0  12.0  
CAUSE.CATEGORY                                               
equipment failure               4.0   4.0   NaN   2.0   4.0  
fuel supply emergency           5.0   3.0   NaN   2.0   4.0  
intentional attack             26.0  21.0  34.0  24.0  24.0  
islanding                       4.0   3.0   3.0   3.0   3.0  
public appeal                  18.0   1.0   2.0   NaN   2.0  
severe weather                 83.0  58.0  62.0  38.0  65.0  
system operability disruption  13.0   4.0   8.0   3.0   9.0

**Interesting Aggregates:**

Pivoting on Cause Category and Month shows us that severe weather is the leading cause of outages and that the month with the most outages is June with the leading cause that month being severe weather.

### Assessment of Missingness

In [12]:
outage_clean.columns[outage_clean.isnull().any()]

Index(['MONTH', 'CLIMATE.REGION', 'ANOMALY.LEVEL', 'CLIMATE.CATEGORY',
       'CAUSE.CATEGORY.DETAIL', 'HURRICANE.NAMES', 'OUTAGE.DURATION',
       'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE',
       'IND.PRICE', 'TOTAL.PRICE', 'RES.SALES', 'COM.SALES', 'IND.SALES',
       'TOTAL.SALES', 'RES.PERCEN', 'COM.PERCEN', 'IND.PERCEN', 'POPDEN_UC',
       'POPDEN_RURAL', 'OUTAGE.START', 'OUTAGE.RESTORATION'],
      dtype='object')

**NMAR Analysis:**

RES.PRICE, COM.PRICE, IND.PRICE, and TOTAL.PRICE could be NMAR because sectors that charge unreasonable prices for electricity will not want to report. A coloumn we could add that could make it MAR is utiltiy to income proportion so we can see where people are paying higher for utitlites while making lower income.

**CAUSE.CATEGORY and CLIMATE.CATEGORY**

In [132]:
outage_mcar = outage_clean.copy()
outage_mcar['CLIMATE_MISSING'] = outage_mcar['CLIMATE.CATEGORY'].isna()
outage_mcar.head()

outage_dist = (
    outage_mcar
    .assign(CLIMATE_MISSING=outage_mcar['CLIMATE.CATEGORY'].isna())
    .pivot_table(index='CAUSE.CATEGORY', columns='CLIMATE_MISSING', aggfunc='size')
)

# Added just to make the resulting pivot table easier to read.
outage_dist.columns = ['CLIMATE_MISSING = False', 'CLIMATE_MISSING = True']

outage_dist = outage_dist / outage_dist.sum()
outage_dist

CLIMATE_MISSING = False  CLIMATE_MISSING = True
CAUSE.CATEGORY                                                                
equipment failure                             0.037377                0.333333
fuel supply emergency                         0.032787                0.111111
intentional attack                            0.274098                     NaN
islanding                                     0.030164                     NaN
public appeal                                 0.045246                     NaN
severe weather                                0.497705                0.444444
system operability disruption                 0.082623                0.111111

In [133]:
outage_dist.plot(kind='barh', title='Cause Category by Missingness of Climate Category', barmode='group')

In [134]:
n_repetitions = 500
shuffled = outage_mcar.copy()

tvds = []
for _ in range(n_repetitions):
    
    # Shuffling months
    shuffled['CAUSE.CATEGORY'] = np.random.permutation(shuffled['CAUSE.CATEGORY'])
    
    # Computing and storing the TVD.
    pivoted = (
        shuffled
        .pivot_table(index='CAUSE.CATEGORY', columns='CLIMATE_MISSING', aggfunc='size')
        .apply(lambda x: x / x.sum())
    )
    
    tvd = pivoted.diff(axis=1).iloc[:, -1].abs().sum() /2
    tvds.append(tvd)
observed_tvd = outage_dist.diff(axis=1).iloc[:, -1].abs().sum() /2
observed_tvd


0.22801457194899818

In [178]:
fig = px.histogram(pd.DataFrame(tvds), x=0, nbins=50, histnorm='probability', 
                   title='Empirical Distribution of the TVD (Climate Category - Cause Category)')
fig.add_vline(x=observed_tvd, line_color='red')
fig.add_annotation(text=f'<span style="color:red">Observed TVD = {round(observed_tvd, 2)}</span>',
                   x=1.5 * observed_tvd, showarrow=False, y=0.16)
fig.update_layout(yaxis_range=[0, 0.2])

In [ ]:
fig.write_html('missingness-MCAR.html', include_plotlyjs='cdn')

In [137]:
np.mean(np.array(tvds) >= observed_tvd)

0.336

**Missingness Dependency:**

*CLIMATE.CATEGORY and CAUSE.CATEGORY*
- We fail to reject the null.
- The null stated that the distribution of 'CAUSE.CATEGORY' when 'CLIMATE.CATEGORY' is missing is the same as the distribution of 'CAUSE.CATEGORY' when 'CLIMATE.CATEGORY' is not missing.
- We conclude that the missingness in the 'CAUSE.CATEGORY' column is not dependent on 'CLIMATE.CATEGORY'.
- Thus MCAR

In [139]:
outage_mar = outage_clean.copy()
outage_mar['CLIMATE_MISSING'] = outage_mar['CLIMATE.CATEGORY'].isna()
outage_mar.head()

outage_dist = (
    outage_mar
    .assign(CLIMATE_MISSING=outage_mar['CLIMATE.CATEGORY'].isna())
    .pivot_table(index='NERC.REGION', columns='CLIMATE_MISSING', aggfunc='size')
)

# Added just to make the resulting pivot table easier to read.
outage_dist.columns = ['CLIMATE_MISSING = False', 'CLIMATE_MISSING = True']

outage_dist = outage_dist / outage_dist.sum()
outage_dist

CLIMATE_MISSING = False  CLIMATE_MISSING = True
NERC.REGION                                                 
ASCC                             NaN                0.111111
ECAR                        0.022295                     NaN
FRCC                        0.028197                0.111111
FRCC, SERC                  0.000656                     NaN
HECO                        0.001967                     NaN
HI                          0.000656                     NaN
MRO                         0.029508                0.111111
NPCC                        0.098361                     NaN
PR                          0.000656                     NaN
RFC                         0.274098                0.111111
SERC                        0.131803                0.444444
SPP                         0.043279                0.111111
TRE                         0.072787                     NaN
WECC                        0.295738                     NaN

In [140]:
outage_dist.plot(kind='barh', title='NERC.REGION by Missingness of Climate Category', barmode='group')

In [141]:
n_repetitions = 500
shuffled = outage_mar.copy()

tvds = []
for _ in range(n_repetitions):
    
    # Shuffling months
    shuffled['NERC.REGION'] = np.random.permutation(shuffled['NERC.REGION'])
    
    # Computing and storing the TVD.
    pivoted = (
        shuffled
        .pivot_table(index='NERC.REGION', columns='CLIMATE_MISSING', aggfunc='size')
        .apply(lambda x: x / x.sum())
    )
    
    tvd = pivoted.diff(axis=1).iloc[:, -1].abs().sum() /2
    tvds.append(tvd)

In [142]:
observed_tvd = outage_dist.diff(axis=1).iloc[:, -1].abs().sum() /2
observed_tvd

0.3539890710382514

In [172]:
fig = px.histogram(pd.DataFrame(tvds), x=0, nbins=50, histnorm='probability', 
                   title='Empirical Distribution of the TVD (Climate Category - NERC Region')
fig.add_vline(x=observed_tvd, line_color='red')
fig.add_annotation(text=f'<span style="color:red">Observed TVD = {round(observed_tvd, 2)}</span>',
                   x=1.5 * observed_tvd, showarrow=False, y=0.16)
fig.update_layout(yaxis_range=[0, 0.2])

In [ ]:
fig.write_html('missingness-MAR.html', include_plotlyjs='cdn')

In [144]:
np.mean(np.array(tvds) >= observed_tvd)

0.034

**Missingness Dependency:**

*CLIMATE.CATEGORY and NERC.REGION*
- We reject the null.
- The null stated that the distribution of 'NERC.REGION' when 'CLIMATE.CATEGORY' is missing is the same as the distribution of 'NERC.REGION' when 'CLIMATE.CATEGORY' is not missing.
- We conclude that the missingness in the 'CLIMATE.CATEGORY' column is not dependent on 'NERC.REGION'.
- Thus MAR

### Hypothesis Testing

In [151]:
outage_clean.groupby(['CAUSE.CATEGORY']).count().idxmax()

OBS                      severe weather
YEAR                     severe weather
MONTH                    severe weather
U.S._STATE               severe weather
POSTAL.CODE              severe weather
NERC.REGION              severe weather
CLIMATE.REGION           severe weather
ANOMALY.LEVEL            severe weather
CLIMATE.CATEGORY         severe weather
CAUSE.CATEGORY.DETAIL    severe weather
HURRICANE.NAMES          severe weather
OUTAGE.DURATION          severe weather
DEMAND.LOSS.MW           severe weather
CUSTOMERS.AFFECTED       severe weather
RES.PRICE                severe weather
COM.PRICE                severe weather
IND.PRICE                severe weather
TOTAL.PRICE              severe weather
RES.SALES                severe weather
COM.SALES                severe weather
IND.SALES                severe weather
TOTAL.SALES              severe weather
RES.PERCEN               severe weather
COM.PERCEN               severe weather
IND.PERCEN               severe weather


**Hypothesis Testing:**

Clearly state a pair of hypotheses and perform a hypothesis test or permutation test that is not related to missingness. Feel free to use the “sample questions” in each of the dataset descriptions or create your own. This should be the question that is stated clearly at the top of your report.

Clearly state your null and alternative hypotheses, your choice of test statistic and significance level, the resulting p-value, and your conclusion. Justify why these choices are good choices for answering the question you are trying to answer.

Optional: Embed a visualization related to your hypothesis test in your website.

Tip: When making writing your conclusions to the statistical tests in this project, never use language that implies an absolute conclusion; since we are performing statistical tests and not randomized controlled trials, we cannot prove that either hypothesis is 100% true or false.


**Null Hypothesis**
- In the population durations of outages caused by 'severe weather' and those with any other cause have the same distribution, and the observed differences in our samples are due to random chance. 'Severe Weather' vs all other cause categories have no relationship to duration of the outage. In other words 'severe weather' and all other cause category labels may well have been assigned at random. Simple: outage durations come from the same distribution.

**Alternative Hypothesis**
- In the population, outages caused by 'severe weather' have longer outage durations than others, on average. The observed difference in our samples cannot be explained by random chance alone. Outages caused by 'severe weather' and those that aren't come from different population distributions. That is, they come from different data generating processes. Simple: outage durations come from different distributions.

**Test Statistic**
- Our test statistic will be *mean outage duration of outages caused by severe weather* − *mean outage duration of outages not caused by severe weather*

**Significance Level**
- Our significance level will be 0.05

**P-Value**
- The p-value is 0

**Conclusion**
- Under the null hypothesis, we rarely see differences as large as 2537 duration. Therefore, we reject the null hypothesis that the two groups come from the same distribution. Thus the observed differences in our samples are not due to random chance. And we can conclude that outages caused by 'severe weather' vs outages not caused by severe weather have some relationship to the duration of the outages.

In [153]:
outage_test = outage_clean.copy()
outage_test['SEVERE.WEATHER'] = outage_test['CAUSE.CATEGORY'] == 'severe weather'


In [154]:
group_means = outage_test.groupby('SEVERE.WEATHER')['OUTAGE.DURATION'].mean()
group_means

SEVERE.WEATHER
False    1346.178962
True     3883.985215
Name: OUTAGE.DURATION, dtype: float64

In [159]:
group_means.loc[True] - group_means.loc[False]

2537.8062533051298

In [161]:
n_repetitions = 500

differences = []
for _ in range(n_repetitions):
    
    # Step 1: Shuffle the weights and store them in a DataFrame.
    with_shuffled = outage_test.assign(Shuffled_Durations=np.random.permutation(outage_test['OUTAGE.DURATION']))

    # Step 2: Compute the test statistic.
    # Remember, alphabetically, False comes before True,
    # so this computes True - False.
    group_means = (
        with_shuffled
        .groupby('SEVERE.WEATHER')
        .mean()
        .loc[:, 'Shuffled_Durations']
    )
    difference = group_means.diff().iloc[-1]
    
    # Step 4: Store the result
    differences.append(difference)
    

In [162]:
observed_difference = outage_test.groupby('SEVERE.WEATHER')['OUTAGE.DURATION'].mean().diff().iloc[-1]
observed_difference

2537.8062533051298

In [168]:
np.mean(differences >= observed_difference)

0.0

In [167]:
fig = px.histogram(pd.DataFrame(differences), x=0, nbins=50, histnorm='probability', 
                   title='Empirical Distribution of the Mean Differences in Outage Durations (Severe Weather - Non-Severe Weather)')
fig.add_vline(x=observed_difference, line_color='red')
fig.update_layout(xaxis_range=[-3000, 3000])

In [ ]:
fig.write_html('permutation-test.html', include_plotlyjs='cdn')